File: src/nssh/cli/ctx/edit.py
"""Context edit command for nssh ctx."""

from __future__ import annotations

from typing import Any, Dict, Optional

from nssh.cli import click
from nssh.cli.common.banner import FAIL, NOOP, OK, banner
from nssh.cli.common.credentials import complete_context, get_manager
from nssh.cli.common.prompt import (
    ask_text,
    prompt_password_with_confirmation,
    prompt_required,
)
from nssh.core.ui.console import get_console

console = get_console()


def _find_context_by_name(cm, name: str) -> Optional[Dict[str, Any]]:
    """Find a context entry by name from list_contexts()."""
    for entry in cm.list_contexts():
        if entry.get("name") == name:
            return entry
    return None


@click.command(short_help="Edit context credentials")
@click.argument("name", required=False, default=None, shell_complete=complete_context)
@click.option(
    "--domain", "-d", default=None, help="Set domain suffix for auto-selection"
)
@click.option("--ssh-config", "-s", default=None, help="Set SSH config include file")
@click.option("--username", "-u", default=None, help="Set fallback username")
@click.option("--password", "-p", default=None, help="Set fallback password")
@click.pass_context
def edit_command(
    ctx: click.Context,
    name: Optional[str],
    domain: Optional[str],
    ssh_config: Optional[str],
    username: Optional[str],
    password: Optional[str],
) -> None:
    """Edit context credentials and settings.

    Sets or replaces the fallback username, password, domain, and SSH config file.
    Use flags for non-interactive updates (empty string clears domain).
    """
    cm = get_manager(ctx)

    with banner("EDIT SSH CONTEXT", OK) as set_outcome:
        _edit_context(cm, name, domain, ssh_config, username, password, set_outcome)


def _edit_context(
    cm, name, domain, ssh_config, username, password, set_outcome
) -> None:
    """Internal implementation for editing a context."""
    # Determine if running non-interactively (any flag provided)
    non_interactive = (
        domain is not None
        or ssh_config is not None
        or username is not None
        or password is not None
    )

    final_name = prompt_required("Context name", name)

    # Check if context exists
    context_entry = _find_context_by_name(cm, final_name)
    if context_entry is None:
        console.print(f"[red]Error: Context '{final_name}' not found[/red]")
        console.print("\nCreate it with:")
        console.print(f"  [cyan]nssh ctx add {final_name}[/cyan]")
        set_outcome(FAIL)
        raise SystemExit(1)

    current_cred = context_entry.get("credential")
    current_domain = context_entry.get("domain", "")
    current_ssh_config = context_entry.get("git_include_file", "")

    # Non-interactive mode: only update what was specified
    if non_interactive:
        # Update domain if specified
        if domain is not None and domain != current_domain:
            try:
                cm.update_context_domain(final_name, domain or None)
                if domain:
                    console.print(f"[green]✓[/green] Domain set to '{domain}'")
                else:
                    console.print("[yellow]![/yellow] [dim]Domain cleared[/dim]")
            except Exception as exc:
                console.print(f"[red]Error updating domain: {exc}[/red]")
                set_outcome(FAIL)
                raise SystemExit(1)

        # Update SSH config file if specified
        if ssh_config is not None and ssh_config != current_ssh_config:
            if not ssh_config:
                console.print("[red]Error: SSH config file cannot be empty[/red]")
                set_outcome(FAIL)
                raise SystemExit(1)
            try:
                cm.update_context_include_file(final_name, ssh_config)
                console.print(f"SSH config set to '{ssh_config}'")
            except Exception as exc:
                console.print(f"[red]Error updating SSH config: {exc}[/red]")
                set_outcome(FAIL)
                raise SystemExit(1)

        # Update credential if username or password specified
        if username is not None or password is not None:
            final_username = username or (
                current_cred["username"] if current_cred else None
            )
            final_password = password

            if not final_username:
                console.print("[red]Error: Username required[/red]")
                set_outcome(FAIL)
                raise SystemExit(1)
            if not final_password:
                console.print(
                    "[red]Error: Password required when setting credentials[/red]"
                )
                set_outcome(FAIL)
                raise SystemExit(1)

            try:
                cm.add_context_credential(
                    final_name, final_username, final_password, overwrite=True
                )
                console.print(f"Credential set for '{final_username}'")
            except Exception as exc:
                console.print(f"[red]Error: {exc}[/red]")
                set_outcome(FAIL)
                raise SystemExit(1)

        console.print(f"[green]✓[/green] Context '{final_name}' updated")
        return

    # Interactive mode
    if current_cred:
        console.print(f"[dim]Current username: {current_cred['username']}[/dim]")
    else:
        console.print("[dim]No credential currently set[/dim]")

    if current_domain:
        console.print(f"[dim]Current domain: {current_domain}[/dim]")
    else:
        console.print("[dim]No domain currently set[/dim]")

    if current_ssh_config:
        console.print(f"[dim]Current SSH config: {current_ssh_config}[/dim]")

    console.print()

    changes_made = False

    # SSH config prompt
    final_ssh_config = ask_text(
        "SSH config file",
        default=current_ssh_config,
        allow_empty=True,
    )

    if final_ssh_config != current_ssh_config and final_ssh_config:
        try:
            cm.update_context_include_file(final_name, final_ssh_config)
            console.print(f"[green]✓[/green] SSH config set to '{final_ssh_config}'")
            changes_made = True
        except Exception as exc:
            console.print(f"[red]Error updating SSH config: {exc}[/red]")
            set_outcome(FAIL)
            raise SystemExit(1)

    # Domain prompt (- to clear)
    final_domain = ask_text(
        "Domain suffix (- to clear)",
        default=current_domain or "",
        allow_empty=True,
    )
    if final_domain == "-":
        final_domain = ""

    if final_domain != current_domain:
        try:
            cm.update_context_domain(final_name, final_domain or None)
            changes_made = True
            if final_domain:
                console.print(f"[green]✓[/green] Domain set to '{final_domain}'")
            else:
                console.print("[yellow]![/yellow] [dim]Domain cleared[/dim]")
        except Exception as exc:
            console.print(f"[red]Error updating domain: {exc}[/red]")
            set_outcome(FAIL)
            raise SystemExit(1)

    # Username prompt (- to skip)
    default_username = current_cred["username"] if current_cred else ""
    final_username = ask_text(
        "Username (- to skip)",
        default=default_username,
        allow_empty=True,
    )
    if final_username == "-":
        final_username = ""

    if not final_username:
        if not changes_made:
            set_outcome(NOOP)
        return

    # Password prompt - empty keeps existing if username unchanged
    try:
        final_password = prompt_password_with_confirmation(
            f"Password for {final_username}",
            allow_empty=True,
        )
    except ValueError as exc:
        console.print(f"[red]{exc}[/red]")
        set_outcome(FAIL)
        raise SystemExit(1)

    # Skip credential update if password empty and username unchanged
    if not final_password:
        if final_username == default_username and current_cred:
            console.print("[dim]Keeping existing credential[/dim]")
            if not changes_made:
                set_outcome(NOOP)
            return
        else:
            console.print("[red]Error: Password required for new username[/red]")
            set_outcome(FAIL)
            raise SystemExit(1)

    try:
        cm.add_context_credential(
            final_name, final_username, final_password, overwrite=True
        )
        console.print(f"[green]✓[/green] Credential updated for '{final_username}'")
    except Exception as exc:
        console.print(f"[red]Error: {exc}[/red]")
        set_outcome(FAIL)
        raise SystemExit(1)

File: internal/cli/ctx/edit.go
package ctx

import (
	"fmt"
	"strings"

	clisession "github.com/ntwrknrd/nssh/internal/cli/session"
	"github.com/ntwrknrd/nssh/internal/ui"
	"github.com/ntwrknrd/nssh/internal/vault"
	"github.com/spf13/cobra"
)

// NewEditCmd creates the ctx edit command.
func NewEditCmd() *cobra.Command {
	var (
		domain    string
		sshConfig string
		username  string
	)

	cmd := &cobra.Command{
		Use:   "edit [name]",
		Short: "Edit context settings",
		Long: `Edit context credentials and settings.

Sets or replaces the fallback username, domain, and SSH config file.
Use flags for non-interactive updates (empty string clears domain).`,
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			var name string
			if len(args) > 0 {
				name = args[0]
			}

			// Determine if any flags were explicitly set
			domainSet := cmd.Flags().Changed("domain")
			sshConfigSet := cmd.Flags().Changed("ssh-config")
			usernameSet := cmd.Flags().Changed("username")

			return runEdit(name, domain, sshConfig, username,
				domainSet, sshConfigSet, usernameSet)
		},
	}

	cmd.Flags().StringVarP(&domain, "domain", "d", "", "set domain suffix for auto-selection")
	cmd.Flags().StringVarP(&sshConfig, "ssh-config", "s", "", "set SSH config include file")
	cmd.Flags().StringVarP(&username, "username", "u", "", "set fallback username (password prompted securely)")

	return cmd
}

func runEdit(name, domain, sshConfig, username string,
	domainSet, sshConfigSet, usernameSet bool) error {

	mgr, err := clisession.NewManager(vault.Auto())
	if err != nil {
		return err
	}

	// Unlock vault if needed and TTY is available
	_ = clisession.TryUnlockIfTTY(mgr)

	// Interactive mode if no flags provided
	nonInteractive := domainSet || sshConfigSet || usernameSet

	// Show banner first
	ui.CommandStart("EDIT CONTEXT")

	// Get context name
	var finalName string
	if name != "" {
		finalName = name
	} else {
		// List available contexts for selection
		contexts, err := mgr.ListContexts()
		if err != nil {
			ui.Error("Failed to list contexts: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		if len(contexts) == 0 {
			ui.Warning("No contexts configured")
			ui.Info("Create one with: nssh ctx add")
			ui.CommandEnd(ui.StatusWarning)
			return nil
		}

		// Build options for select
		options := make([]ui.SelectOption, len(contexts))
		for i, ctx := range contexts {
			label := ctx.Name
			if ctx.Domain != "" {
				label = fmt.Sprintf("%s (%s)", ctx.Name, ctx.Domain)
			}
			options[i] = ui.SelectOption{Label: label, Value: ctx.Name}
		}

		finalName, err = ui.Select("Select context", options)
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}
	}

	// Get context
	ctx, err := mgr.GetContext(finalName)
	if err != nil {
		ui.Error("Failed to get context: %s", err)
		ui.CommandEnd(ui.StatusError)
		return nil
	}
	if ctx == nil {
		ui.Error("Context '%s' not found", finalName)
		ui.Info("Create it with: nssh ctx add %s", finalName)
		ui.CommandEnd(ui.StatusError)
		return nil
	}

	if nonInteractive {
		return runEditNonInteractive(mgr, finalName, ctx, domain, sshConfig, username,
			domainSet, sshConfigSet, usernameSet)
	}

	return runEditInteractive(mgr, finalName, ctx)
}

func runEditNonInteractive(mgr *vault.Manager, name string, ctx *vault.ContextEntry,
	domain, sshConfig, username string,
	domainSet, sshConfigSet, usernameSet bool) error {

	// Update domain if specified
	if domainSet && domain != ctx.Domain {
		if err := mgr.UpdateContextDomain(name, domain); err != nil {
			ui.Error("Failed to update domain: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		if domain != "" {
			ui.Success("Domain set to '%s'", domain)
		} else {
			ui.Warning("Domain cleared")
		}
	}

	// Update SSH config file if specified
	if sshConfigSet && sshConfig != ctx.GitIncludeFile {
		if sshConfig == "" {
			ui.Error("SSH config file cannot be empty")
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		if err := mgr.UpdateContextIncludeFile(name, sshConfig); err != nil {
			ui.Error("Failed to update SSH config: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		ui.Success("SSH config set to '%s'", sshConfig)
	}

	// Update credential if username specified
	if usernameSet {
		if username == "" {
			ui.Error("Username cannot be empty")
			ui.CommandEnd(ui.StatusError)
			return nil
		}

		// Prompt for password securely
		password, err := ui.PasswordWithConfirm("Password")
		if err != nil {
			ui.Error("Password error: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}

		sec := newSecretFromString(password)
		if err := mgr.AddContextCredential(name, username, sec, true); err != nil {
			ui.Error("Failed to update credential: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		ui.Success("Credential set for '%s'", username)
	}

	ui.Success("Context '%s' updated", name)
	ui.CommandEnd(ui.StatusSuccess)
	return nil
}

func runEditInteractive(mgr *vault.Manager, name string, ctx *vault.ContextEntry) error {
	// Show current configuration (skip newline - CommandStart already added one)
	ui.SubSection("Current Configuration", true)
	showContextConfig(ctx)

	// Main menu
	menuOptions := []ui.SelectOption{
		{Label: "Edit SSH config path", Value: "ssh_config"},
		{Label: "Edit domain", Value: "domain"},
		{Label: "Edit credential", Value: "credential"},
		{Label: "Cancel", Value: "cancel"},
	}

	choice, err := ui.Select("Action", menuOptions)
	if err != nil {
		ui.Abort("Canceled")
		ui.CommandEnd(ui.StatusAbort)
		return nil
	}

	switch choice {
	case "ssh_config":
		// Get files used by other contexts (exclude current)
		contexts, _ := mgr.ListContexts()
		usedFiles := make([]string, 0, len(contexts))
		for _, c := range contexts {
			if c.Name != name {
				usedFiles = append(usedFiles, c.GitIncludeFile)
			}
		}

		newConfig, err := selectSSHConfigFile(ctx.GitIncludeFile, usedFiles...)
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}
		if newConfig != ctx.GitIncludeFile {
			if err := mgr.UpdateContextIncludeFile(name, newConfig); err != nil {
				ui.Error("Failed to update SSH config: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			ui.Success("SSH config set to '%s'", newConfig)
			ui.CommandEnd(ui.StatusSuccess)
		} else {
			ui.Noop("No changes made")
			ui.CommandEnd(ui.StatusNoop)
		}

	case "domain":
		editOptions := []ui.SelectOption{
			{Label: "Set domain", Value: "set"},
		}
		if ctx.Domain != "" {
			editOptions = append(editOptions, ui.SelectOption{Label: "Clear domain", Value: "clear"})
		}
		editOptions = append(editOptions, ui.SelectOption{Label: "Back", Value: "back"})

		editChoice, err := ui.Select("Domain option", editOptions)
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}

		switch editChoice {
		case "set":
			defaultVal := ""
			if ctx.Domain != "" {
				defaultVal = ctx.Domain
			}
			newDomain, err := ui.InputWithDefault("Domain", defaultVal)
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if newDomain != "" && newDomain != ctx.Domain {
				if err := mgr.UpdateContextDomain(name, newDomain); err != nil {
					ui.Error("Failed to update domain: %s", err)
					ui.CommandEnd(ui.StatusError)
					return nil
				}
				ui.Success("Domain set to '%s'", newDomain)
				ui.CommandEnd(ui.StatusSuccess)
			} else {
				ui.Noop("No changes made")
				ui.CommandEnd(ui.StatusNoop)
			}
		case "clear":
			if err := mgr.UpdateContextDomain(name, ""); err != nil {
				ui.Error("Failed to clear domain: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			ui.Success("Domain cleared")
			ui.CommandEnd(ui.StatusSuccess)
		case "back":
			ui.Noop("No changes made")
			ui.CommandEnd(ui.StatusNoop)
		}

	case "credential":
		editOptions := []ui.SelectOption{}
		if ctx.Credential != nil {
			editOptions = append(editOptions,
				ui.SelectOption{Label: "Edit username", Value: "username"},
				ui.SelectOption{Label: "Edit password", Value: "password"},
				ui.SelectOption{Label: "Remove credential", Value: "remove"},
			)
		} else {
			editOptions = append(editOptions,
				ui.SelectOption{Label: "Add credential", Value: "add"},
			)
		}
		editOptions = append(editOptions, ui.SelectOption{Label: "Back", Value: "back"})

		editChoice, err := ui.Select("Credential option", editOptions)
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}

		switch editChoice {
		case "add":
			username, err := ui.Input("Username", "")
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if username == "" {
				ui.Warning("Username cannot be empty")
				ui.CommandEnd(ui.StatusWarning)
				return nil
			}
			password, err := ui.PasswordWithConfirm("Password")
			if err != nil {
				ui.Error("Password error: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			sec := newSecretFromString(password)
			if err := mgr.AddContextCredential(name, username, sec, true); err != nil {
				ui.Error("Failed to add credential: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			ui.Success("Credential added for '%s'", username)
			ui.CommandEnd(ui.StatusSuccess)

		case "username":
			newUsername, err := ui.InputWithDefault("Username", ctx.Credential.Username)
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if newUsername == "" {
				ui.Warning("Username cannot be empty")
				ui.CommandEnd(ui.StatusWarning)
				return nil
			}
			if newUsername != ctx.Credential.Username {
				// Keep existing password, just update username
				sec := newSecretFromString(ctx.Credential.Password)
				if err := mgr.AddContextCredential(name, newUsername, sec, true); err != nil {
					ui.Error("Failed to update username: %s", err)
					ui.CommandEnd(ui.StatusError)
					return nil
				}
				ui.Success("Username updated to '%s'", newUsername)
				ui.CommandEnd(ui.StatusSuccess)
			} else {
				ui.Noop("No changes made")
				ui.CommandEnd(ui.StatusNoop)
			}

		case "password":
			password, err := ui.PasswordWithConfirm("New password")
			if err != nil {
				ui.Error("Password error: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			sec := newSecretFromString(password)
			if err := mgr.AddContextCredential(name, ctx.Credential.Username, sec, true); err != nil {
				ui.Error("Failed to update password: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			ui.Success("Password updated for '%s'", ctx.Credential.Username)
			ui.CommandEnd(ui.StatusSuccess)

		case "remove":
			confirm, _ := ui.Confirm("Remove credential?", false)
			if !confirm {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if err := mgr.RemoveContextCredential(name); err != nil {
				ui.Error("Failed to remove credential: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
			ui.Success("Credential removed")
			ui.CommandEnd(ui.StatusSuccess)

		case "back":
			ui.Noop("No changes made")
			ui.CommandEnd(ui.StatusNoop)
		}

	case "cancel":
		ui.Noop("No changes made")
		ui.CommandEnd(ui.StatusNoop)
	}

	return nil
}

func showContextConfig(ctx *vault.ContextEntry) {
	var configLines []string
	configLines = append(configLines, fmt.Sprintf("SSH Config:  %s", ctx.GitIncludeFile))
	hostCount := countHostsInFile(ctx.GitIncludeFile)
	if hostCount > 0 {
		configLines = append(configLines, fmt.Sprintf("Hosts:       %d", hostCount))
	} else {
		configLines = append(configLines, fmt.Sprintf("Hosts:       %s", ui.Gray("(none)")))
	}
	if ctx.Domain != "" {
		configLines = append(configLines, fmt.Sprintf("Domain:      %s", ctx.Domain))
	} else {
		configLines = append(configLines, fmt.Sprintf("Domain:      %s", ui.Gray("(not set)")))
	}
	if ctx.Credential != nil {
		configLines = append(configLines, fmt.Sprintf("Credential:  %s", ctx.Credential.Username))
	} else {
		configLines = append(configLines, fmt.Sprintf("Credential:  %s", ui.Gray("(not set)")))
	}
	ui.Box("", strings.Join(configLines, "\n"))
}
